Service scope implementation

jaellio opened the pull request:
**Please provide a description of this PR:**

TODO:
- [ ] update profile ambient
- [ ] use shared resource types with ambient enablement api
- [ ] more scope testing
- [ ] test default profile impact

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

jaellio committed:
Initial changes to reach label selector from mesh config to update endpoints

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Pass MeshConfig to GlobalMergedWorkloadServicesCollection

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Add cluster and listener service filtering for waypoints and
e/w gateways

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Respond to PR feedback

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Add more tests

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Fix incorrect waypoint logic. In ambient mode, e/w gateways and waypoints
are both node.Type Waypoints

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Remove unused ServicesForGateway

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Updated failing lint and add a release note

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

keithmattix commented on pilot/pkg/model/push_context.go:
```diff
@@ -867,6 +867,7 @@ func (ps *PushContext) ExtraWaypointServices(proxy *Proxy, patches *MergedEnvoyF
 }
 
 // GatewayServices returns the set of services which are referred from the proxy gateways.
+// TODO(jaellio): Do I need to filter services here? 
```
I think this is for old-school gateways; not sure what we decided about filtering those with service scope since they have their own mechanism

keithmattix commented on pilot/pkg/model/push_context.go:
```diff
@@ -958,6 +959,7 @@ const addHostsFromMeshConfigProvidersHandled = 14
 // 1. MeshConfig.ExtensionProviders
 // 2. RequestAuthentication.JwtRules.JwksUri
 // 3. EnvoyFilters with explicitly annotated references
+// TODO(jaellio): Should all of these services have a local scope?
```
Almost definitely yes

keithmattix commented on pilot/pkg/model/push_context.go:
```diff
@@ -2592,3 +2594,10 @@ func (ps *PushContext) ServicesForWaypoint(key WaypointKey) []ServiceInfo {
 }
+
+
+// ServicesForWaypoint returns all services associated with a given waypoint identified by it's WaypointKey
```
Fix comment

keithmattix commented on pilot/pkg/model/service.go:
```diff
@@ -950,6 +950,7 @@ type AmbientIndexes interface {
 	Policies(requested sets.Set[ConfigKey]) []WorkloadAuthorization
 	ServicesForWaypoint(WaypointKey) []ServiceInfo
 	WorkloadsForWaypoint(WaypointKey) []WorkloadInfo
+	ServicesForGateway() []ServiceInfo
```
Why can't we use ServicesForWaypoint with a special waypointkey (maybe an additional field on waypointkey)?

jaellio replied:
Added an `isGateway` bool to the key for now. Might be a better way to do this

keithmattix commented on pilot/pkg/model/service.go:
```diff
@@ -1228,12 +1235,28 @@ func serviceResourceName(s *workloadapi.Service) string {
+	// Global ServiceScope specifies that istiod will automatically expose the matching services' endpoints at the
+	// cluster's east/west gateway. Istio will also automatically share globally matching endpoints with the cluster's
+	// local dataplance that in the local and remote clusters.
+	Global ServiceScope = "global"
```
nit: maybe prefer all caps here?

keithmattix commented on pilot/pkg/model/service.go:
```diff
@@ -1228,12 +1235,28 @@ func serviceResourceName(s *workloadapi.Service) string {
 	// Source is the type that introduced this workload.
 	Source kind.Kind
+	// ScopeForService service key to ServiceScope. This mapping is required since a workload can be a part of
+	// multiple service. A single scope from a service is not sufficiently representive
```
Not sure the last sentence is strictly necessary
```suggestion
	// multiple services.
```

keithmattix commented on pilot/pkg/networking/core/cluster.go:
```diff
@@ -257,8 +257,10 @@ func (configgen *ConfigGeneratorImpl) buildClusters(proxy *model.Proxy, req *mod
 		clusters = append(clusters, configgen.buildWaypointInboundClusters(cb, proxy, req.Push, wps.services)...)
 		clusters = append(clusters, inboundPatcher.insertedClusters()...)
 	default: // Gateways
+		// Get the global services for the gateway
```
If this is for the ambient e/w gateway, this needs to be in the waypoint section. Ambient e/w gateways are `model.Waypoint`s with special sauce. I wouldn't expect this section to change at all

keithmattix commented on pilot/pkg/networking/core/gateway.go:
```diff
@@ -1091,3 +1091,31 @@ func isGatewayMatch(gateway string, gatewayNames []string) bool {
+// findGatewayResources returns workloads and services associated with the gateway proxy
+// TODO(jaellio): Add support for workloads
+// TODO(jaellio): Don't return waypointServer, only need orderedServices or create a new type
+func findGatewayResources(node *model.Proxy, push *model.PushContext) ([]model.WorkloadInfo, *waypointServices) {
```
This should be constrained to the waypoint code. Remember that ambient e/w gateways are waypoints

keithmattix commented on pilot/pkg/networking/core/gateway.go:
```diff
@@ -1091,3 +1091,31 @@ func isGatewayMatch(gateway string, gatewayNames []string) bool {
+		hostName := host.Name(s.Service.Hostname)
+		svc, ok := push.ServiceIndex.HostnameAndNamespace[hostName][s.Service.Namespace]
+		if !ok {
+			continue
```
What does it mean if we get here? Our psuhcontext has the services but the serviceindex doesn't?

keithmattix commented on pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:
```diff
@@ -56,6 +56,7 @@ import (
 type Index interface {
 	Lookup(key string) []model.AddressInfo
 	All() []model.AddressInfo
+	AllGlobalServices() []model.ServiceInfo
```
👍🏾

keithmattix commented on pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:
```diff
@@ -573,29 +574,48 @@ func translateKubernetesCondition(conds []metav1.Condition) map[string]model.Con
+			// TODO(jaellio): I am not sure we need a scopeForService defined on the workload. As far as I know we only lookup workload by service and therefor already have the service scope
+			// we don't need an additional mapping back to service from the workload
+			if w.Workload.ClusterId != string(a.ClusterID) && w.ScopeForService[key] == model.Local {
+				continue
```
A debug log line would be useful as we build this out

keithmattix commented on pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:
```diff
@@ -621,9 +641,31 @@ func (a *index) lookupService(key string) *model.ServiceInfo {
+			res = append(res, wl.AsAddress)
+			continue
+		}
+		// jaellio: Another example of where ScopeForService is not ideal
```
Why is it not ideal?

jaellio replied:
See https://github.com/keithmattix/istio/pull/3/files/5009f6e1503a232c1b16053c55d74b26afc6c180#diff-f8c8b39f11196faa8f4e274b39007bd5fa22b471198ad5da59fdd2fdc9db96fcR614

keithmattix commented on pilot/pkg/serviceregistry/kube/controller/ambient/services.go:
```diff
@@ -197,21 +205,123 @@ func serviceServiceBuilder(
+
+		// TODO(jaellio): When should FetchOne be used? Prior to passing the meshConfig or
+		// when it is needed (in MatchServiceScope)?
+		meshCfg := krt.FetchOne(ctx, meshConfig.AsCollection())
```
Either works as long as this funcs HandlerContext is used to establish the dependency on meshconfig

keithmattix commented on pilot/pkg/serviceregistry/kube/controller/ambient/services.go:
```diff
@@ -197,21 +205,123 @@ func serviceServiceBuilder(
 	}
 }
 
+// TODO(jaellio): does there need to be a lock on reading the meshConfig serviceScopeConfig selectors?
```
Not sure I understand this comment

keithmattix commented on pilot/pkg/serviceregistry/kube/controller/ambient/services.go:
```diff
@@ -197,21 +205,123 @@ func serviceServiceBuilder(
+// There is a filter in discoveryNamespacesFilter that does this, but it is not clear if it is needed here.
+
+// LabelSelectorAsSelector converts a mesh api LabelSelector to a labels.Selector.
+func LabelSelectorAsSelector(ps *meshapi.LabelSelector) (labels.Selector, error) {
```
Is there nothing else that does this already?

jaellio replied:
I borrowed this from Mitch's ambient enablement changes in filter.go. I would have thought so - I'll look

keithmattix commented on pilot/pkg/serviceregistry/kube/controller/ambient/services_test.go:
```diff
@@ -1120,9 +1124,99 @@ func TestServiceConditions(t *testing.T) {
+		svc 	  *v1.Service
+		want model.ServiceScope
+	}{
+		{
```
nit: we probably want more test cases for the final PR to make sure every permutation is caught. One of the comments in the API PR had a good list

keithmattix commented on pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:
```diff
@@ -728,12 +730,19 @@ func workloadEntryWorkloadBuilder(
+		scopeForService := make(map[string]model.ServiceScope)
+		for _, service := range services {
+			// TODO:  Should the scopo come from the service entry? But the service entry will always be local?
+			scopeForService[service.ResourceName()] = model.Local
```
Yeah for workload/serviceenetry, I'm pretty sure the scope of our local instances will always be local

keithmattix commented on pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:
```diff
@@ -884,12 +893,21 @@ func podWorkloadBuilder(
 		w.CanonicalName, w.CanonicalRevision = kubelabels.CanonicalService(p.Labels, w.WorkloadName)
 
+		// populate service to scope mapping
+		// TODO(jaellio): What is the ordering of Service and workload creation? Will service.Scope be populated by the time we get here?
```
I'm pretty sure yes because the workload collection depends on the services collection

keithmattix reviewed (commented):
Good work, this looks about like what I would expect

keithmattix commented on pilot/pkg/networking/core/listener_waypoint.go:
```diff
@@ -85,7 +85,7 @@ func (lb *ListenerBuilder) buildWaypointInbound() []*listener.Listener {
 	var forwarder *listener.Listener
 	if features.EnableAmbientMultiNetwork && isEastWestGateway(lb.node) {
-		wps = nil // TODO: implement service export functionality
+		_, wps = findWaypointResources(lb.node, lb.push)
```
It's now occurring to me that we could probably add e2es for the `TestDirect` codepath for accessing a service through the e/w gateway now. Can you take a stab at building this out? It would be under the ambientmultinetwork conditional [here](https://github.com/istio/istio/blob/0fb4860ebce90d2c5380125f182801d60b6a9d37/tests/integration/ambient/baseline_test.go#L3149). We'd basically want to hit each clusters e/w gateway

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#56654)

my-git9 committed:
support displaying secret info in record message (#56662)

Signed-off-by: xin.li <xin.li@daocloud.io>

zirain committed:
fix clustertrustbundles rbac (#56675)

zirain committed:
fix clusterTrustBundle template (#56676)

my-git9 committed:
support displaying traffic type for waypoint list (#56521)

* support displaying traffic type for waypoint list

Signed-off-by: xin.li <xin.li@daocloud.io>

* add unit test

Signed-off-by: xin.li <xin.li@daocloud.io>

---------

Signed-off-by: xin.li <xin.li@daocloud.io>

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56663)

mike-weiner committed:
Enhance genHTMLFragment to Include Anchor Tags in Table Output (#56645)

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#56683)

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#56684)

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#56688)

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#56693)

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#56694)

hanxiaop committed:
Fix analysis having k8s system namespaces (#56297)

* fix analysis having k8s system namespaces

* add releasenote

istio-testing committed:
Update BASE_VERSION to master-2025-06-19T19-00-49 (#56699)

jaellio committed:
Filter services in ServiceServiceBuilder and check feature
flag for ambient multi network.

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#56708)

ramaraochavali committed:
minor changes to krt (#56647)

* wip

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* fix recomputeMu docs

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

---------

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56685)

jaellio committed:
Move feature flag set to avoid race

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

jaellio committed:
Fix race

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

istio-testing committed:
Automator: update go-control-plane in istio/istio@master (#56724)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56725)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56726)

AritraDey-Dev committed:
support ClusterTrustBundle to v1beta1 API (#56727)

* refactor: migrate ClusterTrustBundle to v1beta1 API

Fixes #56306

Signed-off-by: Aritra Dey <adey01027@gmail.com>

* added release notes

Signed-off-by: Aritra Dey <adey01027@gmail.com>

* modified metadata.yaml

Signed-off-by: Aritra Dey <adey01027@gmail.com>

---------

Signed-off-by: Aritra Dey <adey01027@gmail.com>

adam-cattermole committed:
wasmplugin: Allow configuration of fail reload strategy (#55604)

Signed-off-by: Adam Cattermole <acatterm@redhat.com>

darth-drew committed:
Fix prepetual out of sync status within argocd (#56695)

* Update daemonset.yaml

https://github.com/istio/istio/pull/55955 for reference. Fixes out of sync status within argocd

* change GOMEMLIMIT and GOMAXPROCS divisor from 0 to 1

* Add release notes

jewertow committed:
tls: add PQC compliance policy (#56331)

* tls: add PQC compliance policy

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

* Remove max TLS when compliance policy is PQC

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

* Set cipher suites explicitly in PQC mode

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

* Update documentation for 'pqc' compliance policy

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

* Add a release note

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

---------

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#56744)

Stevenjin8 committed:
Remove unused multiCluster values (#56736)

nicole-lihui committed:
revert istio-remote service when profil=remote (#56524)

* revert istio-remote service when profil=remote

closed #56328

* fix helm values

* fix release note

* fix not install istiod

* iop add EnabledLocalInjectorIstiod & fix configCluster & add ut

* change release note bug to feature

AritraDey-Dev committed:
feat: better istioctl ps output (#56441)

* feat: better istioctl ps output

* feat(proxystatus): dynamically generate XDS sync columns from discovered typeURLs for adaptive proxy-status output

* fix: unit tests

* fix lint target

* fix: release notes

* add a verbosity flag

* fix: release note

* fix: failing tests

* fix: relese note tests

* replace joinWithTabs with strings.Join

Signed-off-by: Aritra Dey <adey01027@gmail.com>

* better istioctl ps output

Signed-off-by: Aritra Dey <adey01027@gmail.com>

---------

Signed-off-by: Aritra Dey <adey01027@gmail.com>

howardjohn committed:
Initial support for ListenerSet (#55595)

* Initial ListenerSet support

* lint

* gen

* rebase

* Rebase

* rebase

* misc fixes

rudrakhp committed:
ensure consistency in inject label validation (#56737)

Signed-off-by: Rudrakh Panigrahi <rudrakh97@gmail.com>

ilrudie committed:
CNI Pod delete edge case (#56738)

* don't leave a dangling UID in the snapshot if we fail with a non-retryable

Signed-off-by: Ian Rudie <ian.rudie@solo.io>

* handle delete edge case

Signed-off-by: Ian Rudie <ian.rudie@solo.io>

* release note

Signed-off-by: Ian Rudie <ian.rudie@solo.io>

---------

Signed-off-by: Ian Rudie <ian.rudie@solo.io>

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56732)

DerekTBrown committed:
Add support for OPTIONAL_MUTUAL to Gateway API (#56742)

* implement OPTIONAL_MUTUAL conversion

* fmt

jaellio committed:
Remove ScopeByService on workload

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

keithmattix committed:
Multicluster AmbientIndex (#56395)

* Build global ambientindex implementation

Use nested join collections to store global versions of each
data structure that underpins the ambient index. We also create indexes
of these collections of collections based on the collection metadata so
that when can query, e.g. the Services from a particular cluster. This
commit also contains the first pass of the merging logic.

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Refactor multicluster collections

krt transformations can be called multiple times for the same inputm and
we were previously initializing new collections in every invocation.
Move to a common implementation of multicluster collections (based on
local and remote sub-collections) and a collection cache that ensures
only 1 collection per remote cluster exists in the global collection.

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Fix nestedjoin collection's seenFirstAddForKey

Previously, we used the collection as the synchronization
mechanism ensuring that we only get 1 add event if a key
gets added to multiple inner nested join collections at
the same time. This doesn't work because the krt event
handling framework has their own handlers and each handler
has its own dedicated goroutine, meaning that using the
collection to synchronize is too broad. Instead, in this
commit, we create a map + mutex data structure per handler,
preventing duplicate adds while ensuring that each handler's
view of the world is correct

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Add multicluster test cases to all existing ambient tests

Confirmed that these tests are all passing as of this commit. Had to
modify the ServiceEntry test slightly to deal with some ordering issues
where the workload entry would return first in the lookup

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Augment merge implementation with real olds

Update the nestedjoin merge implementation so that we can get real old
values. This prevents unnecessary xDS pushes up the stack

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Switch to mergedCache implementation

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Go back to setting collection metadata explicitly

Doing it in the krt options builder looks prettier, but it can lead to
subtle and widely impactful bugs where the builder's metadata is
unintentionally passed through to other clusters' collections.

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Always panic if cluster ID is not set in collection metadata

This will ensure we catch structural issues at compile/test time and we
don't run into tricky, silent bugs

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Check for config clustetr

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Become more robust against crashes

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Fixup test templates

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Use gw.Status for addresses

* Tests use gw.Status instead of Spec

* Add gatewayclass suffixe to default EW gateway SA

* Use status in AmbientSync test

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Fix lint

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* No, actually: become more resilient to crashes

Fix a rare data race that occurs when a cluster gets added to the
clusters collection before it's initialized. Also address several PR
comments.

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Regen golden files

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Fix lint

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

---------

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>
Co-authored-by: Steven Jin Xuan <sjinxuan@microsoft.com>

jaellio committed:
Fix fmt

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

jaellio committed:
Add scope to serviceInfo equality check

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

keithmattix commented:
Closing in favor of the istio/istio PR
